### PR TITLE
central game loop

### PR DESCRIPTION
**Pull request by andrewljohnson**

### CODE REVIEW

**Comment by andrewljohnson** [comment]
@lacker care to make any comments before I finish, I have a bug for sure

File: src/js/gamestate.js [hunk 1]
```diff
@@ -197,18 +208,72 @@ class GameState {
...
+        }
+        let cardTime = Date.now() - card.creationTime
+        card.warm = Math.floor(cardTime % card.attackRate / 1000)
+      }
+    }
+    window.client.forceUpdate()
+
+  }
+
+  // Tick server time for any action that effects game state, like attacks.
```
Comment (lacker) [inline src/js/gamestate.js]: affects

File: src/js/gamestate.js [hunk 2]
```diff
@@ -197,18 +208,72 @@ class GameState {
...
     return true
   }
 
-  // log all moves in the game so far
+  // Do fine-grained time locally, just for things like animations
+  // that don't affect game state.
+  tickLocalTime() {
+    for (let p of this.players) {
+      for (let card of p.board) {
+        if (!card.attackRate) {
```
Comment (lacker) [inline src/js/gamestate.js]: i'm afraid the ! could lead to bugs. could you use a more specific comparator?

File: src/js/gamestate.js [hunk 3]
```diff
@@ -197,18 +208,72 @@ class GameState {
...
+          continue
+        }
+        let cardTime = Date.now() - card.creationTime
+        card.warm = Math.floor(cardTime % card.attackRate / 1000)
```
Comment (lacker) [inline src/js/gamestate.js]: what is this doing?

**Comment by andrewljohnson** [comment]
@lacker check that out, it works

**Pull request closed without merging**